### PR TITLE
feat(datasets) Add `DistributionPartitioner` to Flower Datasets

### PR DESCRIPTION
**Pull request by chongshenng**
## Issue
Some FL researchers use specific NIID distributions such as a pathological power law split but Flower Datasets do not currently support such partitioning schemes.

### Description
The power law splitting procedure are used in several notable research such as `fedprox` and `tamuna`. To ensure that Flower Datasets can be used with Flower baselines, this partitioner needs to be developed.

## Proposal
Add a partitioner that closely reproduce the power-law partitioning scheme.

### Explanation
This partitioner will include the functionalities to:
1. Accept sample counts come from a user-specified distribution.
2. Specify minimum occurrence per label per client.
3. Specify number of labels per client.

Point 1 is intentional so that other distributions (other than a log-normal distribution) can also be flexibly prescribed to this partitioning scheme. It retains its original pathological definition, i.e. partitions only have `num_unique_labels_per_partition`. All samples from the dataset are exhausted during sampling, if the `rescale` parameter is set to `True`.

This implementation is inspired from Li et al. "Federated Optimization in Heterogeneous Networks" (2020) https://arxiv.org/abs/1812.06127.

### Visualizations
The histogram shows the pathological partitioning of a log-normal distribution for 20 partitions with the MNIST dataset, with 5 preassigned number of samples per label, and 2 unique labels per partition:
image

To assess the original power law implementation by Li et al. and the Flower Datasets implementation, we plot and compare the distributions from both implementation below:
image
The plot above uses the original configuration of 1_000 partitions, 5 preassigned number of samples per label, and 2 unique labels per partition. Our Flower Datasets implementation closely matches the original distribution and the $y=x^{-2}$ curve, which validates our implementation.

### Checklist

- [X] Implement proposed change
- [x] Write tests
- [X] Make CI checks pass
- [X] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### CODE REVIEW

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (adam-narozniak) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:37]: Here, we should have either a purely verbal or, if manageable, index-based explanation of what `distribution_array[label_id, index]` means and where such a number of samples will be assigned.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (adam-narozniak) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:54]: Let's reword it so that there's no confusion about `each label`, which can be interpreted as each unique, not each that is meant to be assigned to that partition.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (adam-narozniak) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:58]: Related to this argument - we are not exhausting the samples. The purpose of rescale is (at least in my mind, to use all the samples). If not enabled by default, we should add an argument to allow the use of all samples - counter the float to int rounding.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (adam-narozniak) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:303]: Let's make it more obvious what is the error problem e.g. if the first dim value is not properly set let's explain that the number of unique labels in the dataset is `x` but given `y`. And for the second shape mismatch, e.g., we can defer to the documentation or note that it might be caused by the first shape mismatch in case both are wrong.

File: datasets/flwr_datasets/partitioner/distribution_partitioner_test.py
Comment (adam-narozniak) [inline datasets/flwr_datasets/partitioner/distribution_partitioner_test.py:28]: This can be imported from common, just like in the class implementation file.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (adam-narozniak) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:47]: Let's add the note that currently, it's possible to sample only this number if it is a multiple of the num_unique_labels.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (adam-narozniak) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:49]: This number needs to be checked too. Currently, it works also when the `num_unique_labels_per_client` is bigger than the number of `unique_labels` e.g.  `num_unique_labels_per_client = 20` in the example that you provided.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (adam-narozniak) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:71]: `numpy` import is missing

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (adam-narozniak) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py:54]: Also, when the preassigment is too big the code fails silently (no error but the results do not meet the requirements).  I tried `preassigned_num_samples_per_label = 1_000` (in the example provided)

**Comment by adam-narozniak** [comment]
Let's have tests for all the problems mentioned. Also, it'd be good if we switched to `parameterized` library instead of the `pytest` for consistency.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py:30]: Good point, added!

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py:57]: Good point, added!

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py:74]: Fixed, thanks for picking this up.

File: datasets/flwr_datasets/partitioner/distribution_partitioner_test.py
Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner_test.py:28]: Good catch! Missed this somehow.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py]: Good point, I've added the exhaustive rescaling. For now, it is set to be exhaustive when `rescale = True`. I think it makes sense. I've also updated the PDF above and it's similar, which is good.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (adam-narozniak) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:36]: Let's add a line about the interpretation of the 2-nd dim, which is the number of buckets into which the samples associated with each label will be divided.

**Review by adam-narozniak — COMMENTED** [review]
Left some comments

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py]: Good point, fixed.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py]: > I checked for [1,1] in the described setup below, and it would assigned to the pid=3 (alpha=3; beta=0) but it is to the pid=1 (pid=1 aligned with my interpretation of the code)

Good catch, there's a typo there and should be `(alpha* >= 0 ? ...)`.

The `partition_id` array must first be generated for the i'th row for all `j` and then sorted in ascending order. 

I added the following sentence to be clear:
```
        partition_id = <alpha + beta>
    where,
        <.> denotes the reindexed sequence of partition_ids in monotone increasing
            order for all j's
```

Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py]: I've also added the alternative explanation.

Comment (adam-narozniak) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py]: Sorry, I don't quite get it. I thought of alpha and beta as scalars.

Comment (adam-narozniak) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py]: I'm not quite sure what got changed based on that
```
good catch, there's a typo there and should be (alpha* >= 0 ? ...).
```
Could you explain?

File: datasets/flwr_datasets/partitioner/distribution_partitioner_test.py
Comment (adam-narozniak) [inline datasets/flwr_datasets/partitioner/distribution_partitioner_test.py:101]: If all the methods use the same parameters we can use `parameterized_class` instead of ` @parameterized.expand(`. See https://github.com/adap/flower/blob/main/datasets/flwr_datasets/federated_dataset_test.py

**Review by adam-narozniak — APPROVED** [review]
lgtm!

File: datasets/flwr_datasets/partitioner/distribution_partitioner_test.py
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/distribution_partitioner_test.py:17]:
```suggestion


import unittest
```

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:67]: I'm thinking we should also allow for this to be a list of lists. We'd convert it into a numpy array in `_check_distribution_array_shape_if_needed()` if it's not one already. What do you think?

Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py:67]: @jafermarq I like that option. How about we extend it in a following PR? Merging this PR will unblock a few of our activities, and giving a list of lists is a nice enhancement that can follow this PR.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:384]: wouldn't it be more accurate to call it `unique_labels_counter`?

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:390]:
```suggestion
                    "exceeds that of the original class label distribution."
```

Comment (jafermarq) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py:390]: i'm not sure if the part of "original class label distribution" is accurate. See previous comment in this method.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:404]:
```suggestion
                    f"The number of partitions ({self._num_partitions}) is not divisible "
                    f"by the number of unique labels ({self._num_unique_labels})."
```

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:399]:
```suggestion
                    "The number of partitions ({self._num_partitions}) needs to be smaller than the number of "
                    "samples in the dataset ({self.dataset.num_rows})."
```

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/distribution_partitioner.py:394]: what does ` dataset is given (in load_partition)` mean?

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py:383]: Sure! That works.

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py:389]: I've reworded it to "... exceeds that of the unique labels counter in the given dataset ..."

File: datasets/flwr_datasets/partitioner/distribution_partitioner.py
Comment (chongshenng) [reply datasets/flwr_datasets/partitioner/distribution_partitioner.py:393]: Poor wording, reworded!

**Review by jafermarq — APPROVED** [review]

**Pull request merged**
